src/server/game/Chat/Hyperlinks.cpp

```diff
@@ -0,0 +1,334 @@
+
+#include "Hyperlinks.h"
+#include "advstd.h"
+#include "Common.h"
+#include "DBCStores.h"
+#include "Errors.h"
+#include "SpellInfo.h"
+#include "SpellMgr.h"
+#include "QuestDef.h"
+#include "World.h"
+
+using namespace Trinity::Hyperlinks;
+
+inline bool isHex(char c) { return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f'); }
+// Validates a single hyperlink
+HyperlinkInfo Trinity::Hyperlinks::ParseHyperlink(char const* pos)
+{
+    //color tag
+    if (*(pos++) != '|' || *(pos++) != 'c')
+        return nullptr;
+    for (uint8 i = 0; i < 8; ++i)
+        if (!isHex(*(pos++))) // make sure we don't overrun a terminator
+            return nullptr;
+    // link data start tag
+    if (*(pos++) != '|' || *(pos++) != 'H')
+        return nullptr;
+    // link tag, find next : or |
+    char const* tagStart = pos;
+    size_t tagLength = 0;
+    while (*pos && *pos != '|' && *(pos++) != ':') // we only advance pointer to one past if the last thing is : (not for |), this is intentional!
+        ++tagLength;
+    // ok, link data, skip to next |
+    char const* dataStart = pos;
+    size_t dataLength = 0;
+    while (*pos && *(pos++) != '|')
+        ++dataLength;
+    // ok, next should be link data end tag...
+    if (*(pos++) != 'h')
+        return nullptr;
+    // then visible link text, starts with [
+    if (*(pos++) != '[')
+        return nullptr;
+    // skip until we hit the next ], abort on unexpected |
+    char const* textStart = pos;
+    size_t textLength = 0;
+    while (*pos)
+    {
+        if (*pos == '|')
+            return nullptr;
+        if (*(pos++) == ']')
+            break;
+        ++textLength;
+    }
+    // link end tag
+    if (*(pos++) != '|' || *(pos++) != 'h' || *(pos++) != '|' || *(pos++) != 'r')
+        return nullptr;
+    // ok, valid hyperlink, return info
+    return { pos, tagStart, tagLength, dataStart, dataLength, textStart, textLength };
+}
+
+template <typename T>
+struct LinkTextValidator
+{
+    static bool IsValid(typename T::value_type, char const*, size_t) { return true; }
+};
+
+// str1 is null-terminated, str2 is length-terminated, check if they are exactly equal
+static bool equal_with_len(char const* str1, char const* str2, size_t len)
+{
+    if (!*str1)
+        return false;
+    while (len && *str1 && *(str1++) == *(str2++))
+        --len;
+    return !len && !*str1;
+}
+
+template <>
+struct LinkTextValidator<LinkTags::achievement>
+{
+    static bool IsValid(AchievementLinkData const& data, char const* pos, size_t len)
+    {
+        if (!len)
+            return false;
+        for (uint8 i = 0; i < TOTAL_LOCALES; ++i)
+            if (equal_with_len(data.achievement->Title[i], pos, len))
+                return true;
+        return false;
+    }
+};
+
+template <>
+struct LinkTextValidator<LinkTags::item>
+{
+    static bool IsValid(ItemLinkData const& data, char const* pos, size_t len)
+    {
+        ItemLocale const* locale = sObjectMgr->GetItemLocale(data.item->ItemId);
```

Comment:
this will fail if item with this id has no locale data in separate locale table (applies to every link type)

Reply:
👍